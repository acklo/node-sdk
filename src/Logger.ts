import util from "util";
import { LogLevel } from "./ClientConfiguration";
import chalk, { ChalkFunction } from "chalk";

const levels = {
  off: 50,
  error: 40,
  warn: 30,
  info: 20,
  debug: 10,
};

export interface Logger {
  setLevel(level: LogLevel): void;
  error(format: string, ...args: unknown[]): void;
  warn(format: string, ...args: unknown[]): void;
  info(format: string, ...args: unknown[]): void;
  debug(format: string, ...args: unknown[]): void;
}

export class ConsoleLogger implements Logger {
  constructor(private level: LogLevel) {}

  setLevel(level: LogLevel): void {
    this.level = level;
  }

  error(format: string, ...args: unknown[]): void {
    return this.log("error", format, args);
  }

  warn(format: string, ...args: unknown[]): void {
    return this.log("warn", format, args);
  }

  info(format: string, ...args: unknown[]): void {
    return this.log("info", format, args);
  }

  debug(format: string, ...args: unknown[]): void {
    return this.log("debug", format, args);
  }

  private log(level: LogLevel, format: string, args: unknown[]): void {
    if (levels[this.level] > levels[level]) {
      return;
    }
    const pre = color(level)(`[acklo:${level}]`);
    const formatted = util.format(format, ...args);
    let lines = formatted.split("\n");
    const indent = Array(indentSize(lines[0])).fill(" ").join("");
    lines = lines.map(
      (line, idx) => `${pre}${idx === 0 ? "" : indent} ${line}`
    );
    process.stderr.write(lines.join("\n") + "\n");
  }
}

function indentSize(str: string): number {
  const firstSpaces = str.match(/^ */);
  if (firstSpaces === null) {
    return 0;
  }
  return firstSpaces[0].length;
}

function color(level: LogLevel): ChalkFunction {
  switch (level) {
    case "debug":
      return chalk.green;
    case "info":
      return chalk.blue;
    case "error":
      return chalk.red;
    default:
      return chalk.reset;
  }
}

export class NullLogger implements Logger {
  setLevel(): void {
    // no-op
  }
  error(): void {
    // no-op
  }
  warn(): void {
    // no-op
  }
  info(): void {
    // no-op
  }
  debug(): void {
    // no-op
  }
}

export let logger: Logger = new NullLogger();

export function configureGlobalLoggerInstance(logLevel: LogLevel): void {
  logger = new ConsoleLogger(logLevel);
}
