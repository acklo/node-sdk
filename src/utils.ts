import fs from "fs";
import path from "path";

export async function delay(ms: number): Promise<void> {
  return new Promise<void>((resolve) => {
    setTimeout(() => resolve(), ms);
  });
}

function isFile(path: string): boolean {
  return fs.existsSync(path) && !fs.lstatSync(path).isDirectory();
}

// Based on Jest's implementation for finding the jest.config file:
// https://github.com/facebook/jest/blob/master/packages/jest-config/src/resolveConfigPath.ts
export function findFile(
  filename: string | string[],
  pathToResolve: string
): string | false {
  const filenames = Array.isArray(filename) ? filename : [filename];

  const resolvedPath = filenames
    .map((p) => path.resolve(pathToResolve, p))
    .find(isFile);

  if (resolvedPath) {
    return resolvedPath;
  }

  if (pathToResolve === path.dirname(pathToResolve)) {
    return false;
  }

  return findFile(filenames, path.dirname(pathToResolve));
}
