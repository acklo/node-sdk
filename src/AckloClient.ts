/**
 * A client for interacting with acklo. This is the main entrypoint to the acklo SDK.
 *
 * For extended documentation and getting started guides, check out
 * [acklo docs](https://acklo.app/docs).
 *
 * @example
 * ```js
 * // Create a new instance
 * const acklo = new AckloClient({
 *   applicationName: "my-app",
 *   environmentName: "local",
 *   accessToken: "acklo-123"
 * });
 *
 * // Connect to acklo
 * await acklo.connect();
 *
 * // Get a config value
 * acklo.get("config.port");
 * ```
 */
import { ApiClient } from "./ApiClient";
import {
  ClientConfiguration,
  ClientConfigurationProperties,
} from "./ClientConfiguration";
import {
  ConfigTemplateFile,
  ConfigTemplateEventKind,
  ConfigTemplateValues,
  ConfigTemplateValuesRaw,
  ConfigTemplateValueType,
  ConfigTemplateValueUpdates,
} from "./configTemplateFile/ConfigTemplateFile";
import { HeartbeatSender, HeartbeatSenderEvents } from "./HeartbeatSender";
import { ApiInstanceResponse } from "./types";
import { logger, configureGlobalLoggerInstance } from "./Logger";
import Backoff from "backo2";
import { delay } from "./utils";
import { WebSocketConnection } from "./WebSocketConnection";
import { CommandExecutionHandler } from "./handlers/CommandExecutionHandler";
import { AutoTagger } from "./AutoTagger";
import {
  ConnectionError,
  HumanReadableError,
  isHumanReadableError,
} from "./Errors";
import readConfigTemplateFile from "./configTemplateFile/readConfigTemplateFile";

type ClientState = "disconnected" | "connected";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageVersion = require("../package.json").version;

/**
 * An instance of the acklo client, which can be used for connecting to acklo,
 * retrieving remote configuration for your app, and adding listeners for live updates to
 * your app's configuration.
 *
 * For extended documentation and getting started guides, check out
 * [acklo docs](https://acklo.app/docs).
 *
 * @example
 * ```js
 * // Create a new instance
 * const acklo = new AckloClient({
 *   applicationName: "my-app",
 *   environmentName: "local",
 *   accessToken: "acklo-123"
 * });
 *
 * // Connect to acklo
 * await acklo.connect();
 *
 * // Get a config value
 * acklo.get("config.port");
 * ```
 */
export class AckloClient {
  private readonly apiClient: ApiClient;
  private readonly configTemplate: ConfigTemplateFile;
  private heartbeatSender: HeartbeatSender | null = null;
  private clientConfiguration: ClientConfiguration;
  private instance: ApiInstanceResponse | null = null;
  private webSocketConnection: WebSocketConnection | null = null;
  private state: ClientState = "disconnected";

  /**
   * Create an instance of the acklo client.
   *
   * Once the instance has been created you will be able to get configuration values for
   * your application, however these will _not_ be the latest values configured on the acklo
   * dashboard. In order to get the latest values you'll first need to connect to acklo by
   * calling {@link connect} on your instance.
   *
   * Most of the properties used to configure the instance can also be set as environment
   * variables. There are more details on this in the {@link ClientConfigurationProperties} docs.
   *
   * @example
   * ```js
   * // Create the instance
   * const acklo = new AckloClient({
   *   applicationName: "my-app",
   *   environmentName: "local",
   *   accessToken: "acklo-123"
   * });
   *
   * // This will be the default port configured in your acklo configuration file.
   * acklo.get('config.port_number');
   *
   * // Connect to acklo...
   * await acklo.connect()
   *
   * // Now that you've established a connection to acklo, this will be the latest value
   * // for `config.port_number` as configured on the acklo dashboard.
   * acklo.get('config.port_number');
   * ```
   *
   * @param {ClientConfigurationProperties} properties
   * @throws {@link MissingRequiredConfigurationValuesError} when a required configuration property is missing.
   * @throws {@link MissingConfigTemplateFile} when the acklo config template file could not be found.
   * @throws {@link InvalidConfigTemplateFile} when the acklo config template file is invalid.
   */
  constructor(properties: Partial<ClientConfigurationProperties>) {
    this.clientConfiguration = new ClientConfiguration(properties, process.env);
    configureGlobalLoggerInstance(this.clientConfiguration.logLevel);

    this.apiClient = new ApiClient(
      this.clientConfiguration.accessToken,
      this.clientConfiguration.apiBaseUrl
    );

    try {
      const rawConfigTemplate = readConfigTemplateFile();
      this.configTemplate = new ConfigTemplateFile(rawConfigTemplate);
    } catch (err) {
      this.exitWithError(err, "Failed to read config template file");
    }
  }

  /**
   * Establish a connection to acklo.
   *
   * Once the connection has been established you'll be able to get the latest configuration values
   * from the acklo dashboard and set up listeners so you get notified the instant that a configuration
   * value changes.
   *
   * @example
   * ```js
   * // Using async/await
   * const acklo = await new AckloClient({
   *   applicationName: "my-app",
   *   environmentName: "local",
   *   accessToken: "acklo-123"
   * }).connect();
   *
   * // Using promises
   * const acklo =
   *   new AckloClient({
   *     applicationName: "my-app",
   *     environmentName: "local",
   *     accessToken: "acklo-123"
   *   })
   *  .connect()
   *  .then(() => console.log("Connected successfully"))
   *  .catch(err => console.error("Failed to connect", err));
   *  ```
   *
   * @throws {@link ConnectionError} for any connection errors if you have set
   *   the `continueAfterConnectionError` option to `true`.
   * @returns {Promise<AckloClient>}
   * @param {Object} [options]
   * @param {boolean} [options.continueAfterConnectionError = true] - whether to continue execution without throwing
   *   after an error in connecting to acklo has occurred. Default is `true`.
   */
  async connect(
    options: {
      /**
       * Whether to continue execution without throwing after an error in conencting to acklo has occurred.
       * Default is `true`.
       */
      continueAfterConnectionError: boolean;
    } = {
      continueAfterConnectionError: true,
    }
  ): Promise<AckloClient> {
    if (this.state === "connected") {
      return this;
    }

    try {
      const autoTags = this.clientConfiguration.autoTags
        ? new AutoTagger().tags()
        : {};

      this.instance = await this.apiClient.createInstance(
        this.clientConfiguration.applicationName,
        this.clientConfiguration.environmentName,
        this.configTemplate.toJSON(),
        this.configTemplate.rawContent,
        packageVersion,
        { ...autoTags, ...this.clientConfiguration.tags }
      );

      this.configTemplate.updateValues(
        await this.apiClient.getInstanceConfiguration(this.instance.id)
      );

      this.heartbeatSender = new HeartbeatSender(
        this.instance.id,
        this.apiClient,
        this.clientConfiguration.heartbeatInterval
      ).start();

      this.heartbeatSender.events.on(
        HeartbeatSenderEvents.HeartbeatFailed,
        () =>
          this.reconnect().catch(() => {
            throw new HumanReadableError("Failed to reconnect to acklo.");
          })
      );

      this.webSocketConnection = new WebSocketConnection(
        this.clientConfiguration.webSocketBaseUrl
      );
      this.webSocketConnection.addHandler(
        new CommandExecutionHandler(this.configTemplate, this.apiClient)
      );
      await this.webSocketConnection.connect(this.instance.id);

      this.state = "connected";

      logger.info("Connected to acklo!", {
        instanceName: this.getInstanceName(),
        instanceUrl: this.getInstanceUrl(),
        configuration: this.getConfig(),
      });
    } catch (err) {
      const connectionError = new ConnectionError(err);
      if (options.continueAfterConnectionError) {
        logger.error(connectionError.humanReadableMessage);
        logger.debug("%o", err);
      } else {
        this.exitWithError(err, "Failed to connect");
      }
    }

    return this;
  }

  /**
   * Disconnect from acklo.
   *
   * Once you've disconnected you will no longer receive configuration updates
   * from acklo. It's recommended that you call this method when your application is shutting down
   * (e.g. when it receives `SIGTERM`) in order to gracefully close any open sockets the acklo client
   * has established.
   *
   * @example
   * ```js
   * const acklo = await new AckloClient({
   *   applicationName: "my-app",
   *   environmentName: "local",
   *   accessToken: "acklo-123"
   * }).connect();
   *
   * await acklo.disconnect();
   * ```
   *
   * @returns {Promise<AckloClient>}
   */
  async disconnect(): Promise<AckloClient> {
    if (this.state === "disconnected") {
      return this;
    }

    this.heartbeatSender?.stop();
    this.webSocketConnection?.disconnect();

    if (this.instance) {
      this.instance = null;
    }

    this.state = "disconnected";

    return this;
  }

  /**
   * Returns the name of this application instance as displayed on the acklo dashboard.
   *
   * This will only return a value once the client has been successfully connected to acklo by
   * using the {@link connect} method.
   *
   * The instance name is randomly generated by acklo and is a combination of some recognizable
   * words and random characters.
   *
   * @example
   * ```js
   * acklo.getInstanceName();
   * ```
   *
   * Returns:
   *
   * ```js
   * "ins-delicate-feather-qpcw"
   * ```
   *
   * @returns {string | undefined}
   */
  getInstanceName(): string | undefined {
    return this.instance?.name;
  }

  /**
   * Returns the URL at which you can see this instance on the acklo dashboard.
   *
   * This will only return a value once the client has been successfully connected to acklo by
   * using the {@link connect} method.
   *
   * @example
   * ```js
   * acklo.getInstanceUrl()
   * ```
   *
   * Returns:
   *
   * ```js
   * "https://acklo.app/dashboard/applications/my-app/local/instances/a438372d-2dba-46dc-ada2-3ae043e2c8c3"
   * ```
   *
   * @returns {string | undefined}
   */
  getInstanceUrl(): string | undefined {
    return this.instance?.url;
  }

  /**
   * Returns the latest config for your application.
   *
   * This will return the default values specified in your acklo configuration file (`acklo.config.yml`) until
   * the client has been successfully connected to acklo by using the {@link connect} method.
   *
   * @example
   * ```js
   * acklo.getConfig();
   * ```
   *
   * Returns:
   *
   * ```js
   * {
   *   "config.port": 3000,
   *   "config.greeting": "Howdy, partner",
   *   "feature_switch.new_homepage": true
   * }
   * ```
   *
   * @returns {ConfigTemplateValues}
   */
  getConfig(): ConfigTemplateValues {
    return this.configTemplate.values();
  }

  /**
   * Returns the latest config for your application, but unlike {@link getConfig} does not try to coerce
   * the values stored on the acklo backend to native Javascript types (i.e. booleans, numbers, etc). Instead
   * it returns all values as strings and lets you do your own coercion.
   *
   * Unless you have a good reason to use this method, you probably want {@link getConfig} instead.
   *
   * This will return the default values specified in your acklo configuration file (`acklo.config.yml`) until
   * the client has been successfully connected to acklo by using the {@link connect} method.
   *
   * @example
   * ```js
   * acklo.getConfigRaw();
   * ```
   *
   * Returns:
   *
   * ```js
   * {
   *   "config.port": "3000",
   *   "config.greeting": "Howdy, partner",
   *   "feature_switch.new_homepage": "true"
   * }
   * ```
   *
   * @returns {ConfigTemplateValuesRaw}
   */
  getConfigRaw(): ConfigTemplateValuesRaw {
    return this.configTemplate.valuesRaw();
  }

  /**
   * Returns the latest config value for the given property.
   *
   * This will return the default values specified in your acklo configuration file (`acklo.config.yml`) until
   * the client has been successfully connected to acklo by using the {@link connect} method.
   *
   * `undefined` will be returned if you try to get a value for a config property that does not exist.
   *
   * @example
   * ```js
   * acklo.get("config.port");
   * ```
   *
   * Returns:
   *
   * ```js
   * 3000
   * ```
   *
   * @param {string} id - the ID of the config value you want to get (e.g. `"config.port"`)
   * @returns {ConfigTemplateValueType}
   */
  get(id: string): ConfigTemplateValueType {
    const value = this.getConfig()[id];

    if (value === undefined) {
      logger.warn("Tried to get undefined config property", { id });
    }

    return value;
  }

  /**
   * Returns the latest config value for the given property, but unlike {@link get} does not try to coerce
   * the values stored on the acklo backend to native JavaScript types (i.e. booleans, numbers, etc). Instead
   * it returns all values as strings and lets you do your own coercion.
   *
   * Unless you have a good reason to use this method, you probably want {@link get} instead.
   *
   * This will return the default values specified in your acklo configuration file (`acklo.config.yml`) until
   * the client has been successfully connected to acklo by using the {@link connect} method.
   *
   * `undefined` will be returned if you try to get a value for a config property that does not exist.
   *
   * @example
   * ```js
   * acklo.get("config.port");
   * ```
   *
   * Returns:
   *
   * ```js
   * "3000"
   * ```
   *
   * @param id
   */
  getRaw(id: string): string | undefined {
    return this.getConfigRaw()[id];
  }

  /**
   * Registers the given callback to run as soon as an update is made to this application's config. This
   * allows you to react immediately to configuration changes without needing to poll the {@link get} method.
   *
   * @example
   * ```js
   * // Where "config.port" has been updated to 4000.
   * acklo.onConfigUpdate(updates => {
   *   console.log(`my new port is: ${updates["config.port"].newValue}`);
   *   // Prints
   *   // my new port is: 4000
   * });
   * ```
   *
   * @param callback
   */
  onConfigUpdate(
    callback: (updates: ConfigTemplateValueUpdates) => void
  ): void {
    this.configTemplate.events.on(
      ConfigTemplateEventKind.ConfigUpdate,
      callback
    );
  }

  private async reconnect(): Promise<AckloClient> {
    await this.disconnect();

    const backoff = new Backoff({
      min: 500,
      max: 10_000,
      jitter: 200,
      factor: 1.5,
    });

    while (this.state === "disconnected") {
      logger.info("Attempting to reconnect to acklo...");
      await this.connect();
      await delay(backoff.duration());
    }

    return this;
  }

  /**
   * Exits the process because an unrecoverable error has happened.
   *
   * @private
   */
  private exitWithError(error: Error, message: string): void {
    if (isHumanReadableError(error)) {
      logger.error(error.humanReadableMessage);
      logger.debug("Error details", { error });
    } else {
      logger.error(message, { error });
    }

    process.exit(1);
  }
}
