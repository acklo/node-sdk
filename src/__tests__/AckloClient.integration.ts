import mock from "mock-fs";
import { AckloClient } from "../AckloClient";
import fs from "fs";
import { configureGlobalLoggerInstance } from "../Logger";
import { stderr } from "process";
import { mocked } from "ts-jest/utils";

const TOKEN_ENV_VAR = "ACKLO_INTEGRATION_TEST_TOKEN";
const token = process.env[TOKEN_ENV_VAR];
if (!token) {
  throw new Error(`Need ${TOKEN_ENV_VAR} env var set to run integration tests`);
}

describe("acklo (integration)", () => {
  jest.spyOn(stderr, "write").mockImplementation(undefined);
  const stderrMock = mocked(stderr.write);

  beforeAll(() => {
    configureGlobalLoggerInstance("info");
  });

  afterAll(() => {
    configureGlobalLoggerInstance("off");
  });

  describe("with a valid config template", () => {
    beforeEach(() => {
      mock({
        "acklo.config.yml": fs
          .readFileSync(
            "./src/configTemplateFile/__tests__/configTemplateExample.yml"
          )
          .toString("utf-8"),
      });
    });

    afterEach(() => {
      mock.restore();
    });

    it("connects to the acklo api and retrieves config values", async () => {
      const instance = new AckloClient({
        applicationName: "sdk-test-target",
        environmentName: "integration-test",
        accessToken: token,
      });

      expect(instance.getConfig()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "trace",
        "app_config.port": 3000,
        "feature_switches.motd": "Be good to each other.",
        "feature_switches.new_header": false,
      }
    `);

      await expect(instance.connect()).resolves.toBeInstanceOf(AckloClient);

      expect(instance.getConfig()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "trace",
        "app_config.port": 3000,
        "feature_switches.motd": "Good morning!",
        "feature_switches.new_header": false,
      }
    `);

      expect(instance.getConfigRaw()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "trace",
        "app_config.port": "3000",
        "feature_switches.motd": "Good morning!",
        "feature_switches.new_header": "false",
      }
    `);

      expect(instance.get("app_config.log_level")).toMatchInlineSnapshot(
        `"trace"`
      );

      expect(instance.getRaw("app_config.port")).toMatchInlineSnapshot(
        `"3000"`
      );

      expect(instance.getInstanceName()).toMatch(/^ins-\w+-\w+-\w+$/);
      expect(instance.getInstanceUrl()).toMatch(
        /^http.*applications\/sdk-test-target\/integration-test\/instances\/.*$/
      );

      await expect(instance.disconnect()).resolves.toBeInstanceOf(AckloClient);

      const loggedLines = stderrMock.mock.calls.map((c) => c[0]);
      expect(loggedLines).toMatchObject([
        expect.stringContaining("Connected to acklo!"),
      ]);
    });

    it("returns defaults when the api cannot be reached", async () => {
      const instance = new AckloClient({
        applicationName: "sdk-test-target",
        environmentName: "integration-test",
        accessToken: "not-a-valid-token",
      });

      await expect(instance.connect()).resolves.toBe(instance);

      expect(instance.getConfig()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "trace",
        "app_config.port": 3000,
        "feature_switches.motd": "Be good to each other.",
        "feature_switches.new_header": false,
      }
    `);

      await expect(instance.disconnect()).resolves.toBe(instance);

      const loggedLines = stderrMock.mock.calls.map((c) => c[0].toString());
      expect(loggedLines.filter((l) => l.indexOf("acklo:error") > -1))
        .toMatchInlineSnapshot(`
      Array [
        "[acklo:error] An invalid access token has been provided. Please configure your SDK with a valid access token.
      ",
      ]
    `);
    });
  });

  describe("when the config template has an error in it", () => {
    const mockExit = jest.spyOn(process, "exit").mockImplementation();

    beforeAll(() => {
      mock({
        "acklo.config.yml": "not-valid!",
      });
    });

    afterAll(() => {
      mock.restore();
      mockExit.mockRestore();
    });

    it("exits with a friendly error message", () => {
      stderrMock.mockReset();

      new AckloClient({
        applicationName: "sdk-test-target",
        environmentName: "integration-test",
        accessToken: token,
      });

      expect(stderrMock.mock.calls[0]).toMatchSnapshot();
      expect(mockExit).toHaveBeenCalledWith(1);
    });
  });
});
