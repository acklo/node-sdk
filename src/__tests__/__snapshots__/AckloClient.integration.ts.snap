// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acklo (integration) when the config template has an error in it exits with a friendly error message 1`] = `
Array [
  "[acklo:error] The contents of the acklo config template file (acklo.config.yml) could not be parsed.
[acklo:error]     
[acklo:error] Check the ConfigTemplate docs and make sure that your template matches the expected format:
[acklo:error] https://acklo.app/docs/nodejs_sdk/modules/configtemplate.html
",
]
`;
